motion-pipeline/motion_extraction/audio_analysis/audio_tools.py

```diff
@@ -30,34 +38,56 @@ def standardize_bpm_range(bpm_in: float, bpm_min: float = 80.0, bpm_max: float =
     return bpm
 
 
-def calculate_8beat_segments_with_midpoints(bpm: float, beat_offset: float, duration: float) -> t.Iterable[t.List[float]]:
+def calculate_8beat_segments_with_midpoints(bpm: float, beat_offset: float, duration: float) -> t.Iterable[MusicPhrase]:
+    """Calculate musical phrases (based on 8-beat segments), aligned to the beat.
+    The phrases include midpoints that divide the phrase into individual bars.
+    Takes care not to subdivide into bars when the musical phrase is too short.
+
+    Parameters:
+    bpm (float): The tempo of the music in beats per minute.
+    beat_offset (float): The offset of the first beat in seconds.
+    duration (float): The duration of the music in seconds.
+
+    Returns:
+    A generator that yields a MusicPhrase for each 8-beat segment.
+    """
     sec_per_beat = 60 / bpm
     beats_per_bar = 4
     secs_per_bar = sec_per_beat * beats_per_bar
     bars_per_segment = 2
     secs_per_segment = secs_per_bar * bars_per_segment
 
     ## Generate first segment
-    # Edge case: single segment (< 1.5 segments)
+    # Edge case: single segment (< 1.5 segments; < 3 bars)
     if duration < beat_offset + secs_per_segment * 1.5:
 
-        # Generate up to two mid-points
         mid_points = []
+
         if duration > beat_offset + secs_per_bar * 0.75:
+            # Add a midpoint at the end of the first bar
             mid_points.append(beat_offset + secs_per_bar)
         if duration > beat_offset + secs_per_bar * 1.25:
+            # Add a midpoint at the end of the second bar
             mid_points.append(beat_offset + secs_per_segment)
 
-        yield [0., *mid_points, duration]
+        yield MusicPhrase(beat_offset, mid_points, duration)
```

Comment:
note to Selena: the "yield" keyword is basically a way to have a function return a sequence of values over time, called a generator. This has benefits for writing async and interruptable code, but this function is synchronous and just uses the yield keyword to avoid the boilerplate of creating an empty list at the beginning, appending a bunch of items, and returning the list at the end.